File: src/med_exam_toolkit/inspect.py
from __future__ import annotations

import re
import unicodedata
from pathlib import Path


# ── CJK 对齐工具 ──

def _cjk_len(s: str) -> int:
    return sum(2 if unicodedata.east_asian_width(c) in ("W", "F") else 1 for c in s)


def _ljust(s: str, width: int) -> str:
    return s + " " * max(0, width - _cjk_len(s))


def _trunc(s: str, n: int = 40) -> str:
    s = (s or "").replace("\n", " ").strip()
    return s[:n] + "…" if len(s) > n else s


def _print_options(options: list[str]) -> list[str]:
    """渲染选项，避免双重字母前缀"""
    labels = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    lines = []
    for oi, opt in enumerate(options or []):
        opt = opt.strip()
        if re.match(r'^[A-Za-z][.．、]\s*', opt):
            lines.append(f"         {opt}")
        else:
            key = labels[oi] if oi < len(labels) else str(oi + 1)
            lines.append(f"         {key}. {opt}")
    return lines


# ── 主体实现 ──

def run_inspect(
    bank: str,
    password: str | None,
    filter_modes: tuple,
    filter_units: tuple,
    keyword: str,
    has_ai: bool,
    missing: bool,
    limit: int,
    full: bool,
    show_ai: bool,
) -> None:
    import click
    from med_exam_toolkit.bank import load_bank

    questions = load_bank(Path(bank), password)
    W = 72

    print_summary(questions, bank, W)

    results: list[tuple[int, int]] = []
    for qi, q in enumerate(questions):
        if filter_modes and q.mode not in filter_modes:
            continue
        if filter_units and not any(kw in (q.unit or "") for kw in filter_units):
            continue
        if keyword and keyword not in (q.stem or "") and not any(
            keyword in sq.text for sq in q.sub_questions
        ):
            continue
        for si, sq in enumerate(q.sub_questions):
            if has_ai and not (sq.ai_answer or sq.ai_discuss):
                continue
            if missing and (sq.answer or "").strip() and (sq.discuss or "").strip():
                continue
            results.append((qi, si))

    has_filter = any([filter_modes, filter_units, keyword, has_ai, missing])
    if has_filter:
        click.echo(f"\n  🔎 过滤结果：{len(results)} 个小题\n")
    else:
        n_show = "全部" if limit == 0 else str(limit)
        click.echo(f"\n  📋 题目列表（前 {n_show} 个小题）\n")

    show = results if limit == 0 else results[:limit]
    for qi, si in show:
        print_question(questions[qi], questions[qi].sub_questions[si],
                       qi, si, W, full=full, show_ai=show_ai)

    click.echo(f"{'─' * W}")
    if limit and len(results) > limit:
        click.echo(f"  … 还有 {len(results) - limit} 个，用 --limit 0 显示全部")
    click.echo()


def print_summary(questions: list, bank: str, W: int = 72) -> None:
    from collections import Counter
    import click

    total_q  = len(questions)
    total_sq = sum(len(q.sub_questions) for q in questions)
    no_ans   = sum(1 for q in questions for sq in q.sub_questions
                   if not (sq.answer or "").strip())
    no_dis   = sum(1 for q in questions for sq in q.sub_questions
                   if not (sq.discuss or "").strip())
    has_ai_c = sum(1 for q in questions for sq in q.sub_questions
                   if sq.ai_answer or sq.ai_discuss)
    ai_ans_c = sum(1 for q in questions for sq in q.sub_questions
                   if sq.answer_source == "ai")
    ai_dis_c = sum(1 for q in questions for sq in q.sub_questions
                   if sq.discuss_source == "ai")
    mode_cnt = Counter(q.mode for q in questions)
    cls_cnt  = Counter(q.cls  for q in questions)

    click.echo(f"\n{'═' * W}")
    click.echo(f"  📦 题库：{bank}")
    click.echo(f"{'─' * W}")

    lw, nw = 10, 6

    def stat_row(*pairs):
        parts = [f"{_ljust(label + '：', lw + 1)}{str(val):>{nw}}" for label, val in pairs]
        click.echo("  " + "    ".join(parts))

    stat_row(("大题数", total_q),    ("小题数", total_sq))
    stat_row(("缺答案", no_ans),     ("缺解析", no_dis))
    stat_row(("含AI内容", has_ai_c), ("AI答案兜底", ai_ans_c), ("AI解析兜底", ai_dis_c))

    click.echo(f"\n  题型分布：")
    mode_lw = max((_cjk_len(m or "未知") for m in mode_cnt), default=8)
    for m, c in sorted(mode_cnt.items()):
        click.echo(f"    {_ljust(m or '未知', mode_lw)}  {c:>6} 题")

    if len(cls_cnt) > 1:
        click.echo(f"\n  分类分布：")
        cls_lw = max((_cjk_len(c or "未知") for c in cls_cnt), default=10)
        for c, n in sorted(cls_cnt.items(), key=lambda x: -x[1])[:10]:
            click.echo(f"    {_ljust(c or '未知', cls_lw)}  {n:>6} 题")

    click.echo(f"{'═' * W}")


def print_question(q, sq, qi: int, si: int, W: int = 72,
                   *, full: bool, show_ai: bool) -> None:
    import click

    ans_flag = ("✅" if (sq.answer  or "").strip()
                else "🤖" if (sq.ai_answer or "").strip() else "❓")
    dis_flag = ("✅" if (sq.discuss or "").strip()
                else "🤖" if (sq.ai_discuss or "").strip() else "❓")
    ai_tag   = " [AI]" if (sq.ai_answer or sq.ai_discuss) else ""

    click.echo(f"{'─' * W}")
    click.echo(
        f"  [{qi+1}-{si+1}]  {q.mode}  {q.unit or ''}  "
        f"答案:{ans_flag}  解析:{dis_flag}{ai_tag}"
    )

    if q.stem:
        click.echo(f"  【题干】{q.stem if full else _trunc(q.stem, 100)}")

    text = sq.text or "(无题目文本)"
    click.echo(f"  【题目】{text if full else _trunc(text, 100)}")

    for line in _print_options(sq.options):
        click.echo(line)

    eff_ans = sq.eff_answer
    if eff_ans:
        src  = " (AI)" if sq.answer_source == "ai" else ""
        conf = (f"  [置信:{sq.ai_confidence:.2f}  模型:{sq.ai_model}]"
                if sq.answer_source == "ai" else "")
        click.echo(f"  【答案】{eff_ans}{src}{conf}")

    eff_dis = sq.eff_discuss
    if eff_dis:
        src     = " (AI)" if sq.discuss_source == "ai" else ""
        dis_out = eff_dis if full else _trunc(eff_dis, 150)
        click.echo(f"  【解析】{dis_out}{src}")

    if show_ai and (sq.ai_answer or sq.ai_discuss):
        sep = "─" * ((W - 12) // 2)
        click.echo(f"  {sep} AI原始输出 {sep}")
        if sq.ai_answer and sq.ai_answer != sq.eff_answer:
            click.echo(f"  【AI答案】{sq.ai_answer}"
                       f"  [置信:{sq.ai_confidence:.2f}  模型:{sq.ai_model}]")
        if sq.ai_discuss:
            ai_dis = sq.ai_discuss if full else _trunc(sq.ai_discuss, 150)
            same   = sq.ai_discuss.strip() == (sq.discuss or "").strip()
            note   = "（与官方解析相同）" if same else ""
            click.echo(f"  【AI解析】{ai_dis}{note}")

File: cmd/inspect.go
package cmd

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/spf13/cobra"
	"github.com/zyh001/med-exam-kit/internal/bank"
	"github.com/zyh001/med-exam-kit/internal/models"
	"github.com/zyh001/med-exam-kit/internal/stats"
)

var inspectCmd = &cobra.Command{
	Use:   "inspect",
	Short: "查看 .mqb 题库内容，支持过滤与搜索",
	RunE:  runInspect,
}

func init() {
	rootCmd.AddCommand(inspectCmd)
	inspectCmd.Flags().StringSlice("mode", nil, "过滤题型")
	inspectCmd.Flags().StringSlice("unit", nil, "过滤章节关键词")
	inspectCmd.Flags().String("keyword", "", "题干或题目关键词")
	inspectCmd.Flags().Bool("has-ai", false, "只显示含 AI 补全内容的题")
	inspectCmd.Flags().Bool("missing", false, "只显示缺答案或缺解析的题")
	inspectCmd.Flags().Int("limit", 20, "最多显示多少小题（0=全部）")
	inspectCmd.Flags().Bool("full", false, "显示完整解析（默认截断至 150 字）")
	inspectCmd.Flags().Bool("show-ai", false, "同时显示 AI 原始输出")
}

func runInspect(cmd *cobra.Command, args []string) error {
	bankPath, _ := cmd.Root().PersistentFlags().GetString("bank")
	password, _ := cmd.Root().PersistentFlags().GetString("password")
	modes, _ := cmd.Flags().GetStringSlice("mode")
	units, _ := cmd.Flags().GetStringSlice("unit")
	keyword, _ := cmd.Flags().GetString("keyword")
	hasAI, _ := cmd.Flags().GetBool("has-ai")
	missing, _ := cmd.Flags().GetBool("missing")
	limit, _ := cmd.Flags().GetInt("limit")
	full, _ := cmd.Flags().GetBool("full")
	showAI, _ := cmd.Flags().GetBool("show-ai")

	if bankPath == "" {
		return fmt.Errorf("请用 -b 指定题库路径")
	}

	questions, err := bank.LoadBank(bankPath, password)
	if err != nil {
		return err
	}

	W := 72

	// Print summary header
	printInspectSummary(questions, bankPath, W)

	// Collect matching (qi, si) pairs
	type pair struct{ qi, si int }
	var results []pair

	for qi, q := range questions {
		if len(modes) > 0 && !containsAny(q.Mode, modes) {
			continue
		}
		if len(units) > 0 && !containsAnySubstring(q.Unit, units) {
			continue
		}
		if keyword != "" {
			found := strings.Contains(q.Stem, keyword)
			for _, sq := range q.SubQuestions {
				if strings.Contains(sq.Text, keyword) {
					found = true
					break
				}
			}
			if !found {
				continue
			}
		}
		for si, sq := range q.SubQuestions {
			if hasAI && sq.AIAnswer == "" && sq.AIDiscuss == "" {
				continue
			}
			if missing && strings.TrimSpace(sq.Answer) != "" && strings.TrimSpace(sq.Discuss) != "" {
				continue
			}
			results = append(results, pair{qi, si})
		}
	}

	hasFilter := len(modes) > 0 || len(units) > 0 || keyword != "" || hasAI || missing
	if hasFilter {
		fmt.Printf("\n  🔎 过滤结果：%d 个小题\n\n", len(results))
	} else {
		show := "全部"
		if limit > 0 {
			show = fmt.Sprintf("%d", limit)
		}
		fmt.Printf("\n  📋 题目列表（前 %s 个小题）\n\n", show)
	}

	showList := results
	if limit > 0 && len(showList) > limit {
		showList = showList[:limit]
	}

	for _, p := range showList {
		q := questions[p.qi]
		sq := q.SubQuestions[p.si]
		printInspectQuestion(q, &sq, p.qi, p.si, W, full, showAI)
	}

	fmt.Println(strings.Repeat("─", W))
	if limit > 0 && len(results) > limit {
		fmt.Printf("  … 还有 %d 个，用 --limit 0 显示全部\n", len(results)-limit)
	}
	fmt.Println()
	return nil
}

func printInspectSummary(questions []*models.Question, bankPath string, W int) {
	totalQ := len(questions)
	totalSQ := 0
	noAns := 0
	noDis := 0
	hasAIC := 0
	aiAnsC := 0
	aiDisC := 0
	modeCount := map[string]int{}

	for _, q := range questions {
		modeCount[q.Mode]++
		for _, sq := range q.SubQuestions {
			totalSQ++
			if strings.TrimSpace(sq.Answer) == "" {
				noAns++
			}
			if strings.TrimSpace(sq.Discuss) == "" {
				noDis++
			}
			if sq.AIAnswer != "" || sq.AIDiscuss != "" {
				hasAIC++
			}
			if sq.AnswerSource() == "ai" {
				aiAnsC++
			}
			if sq.DiscussSource() == "ai" {
				aiDisC++
			}
		}
	}

	sep := strings.Repeat("═", W)
	line := strings.Repeat("─", W)
	fmt.Printf("\n%s\n", sep)
	fmt.Printf("  📦 题库：%s\n", bankPath)
	fmt.Println(line)
	fmt.Printf("  %-12s%6d    %-12s%6d\n", "大题数：", totalQ, "小题数：", totalSQ)
	fmt.Printf("  %-12s%6d    %-12s%6d\n", "缺答案：", noAns, "缺解析：", noDis)
	fmt.Printf("  %-12s%6d    %-12s%6d    %-14s%6d\n",
		"含AI内容：", hasAIC, "AI答案兜底：", aiAnsC, "AI解析兜底：", aiDisC)

	fmt.Printf("\n  题型分布：\n")
	modeKV := stats.Summarize(questions, false).ByMode
	for _, m := range modeKV {
		label := m.Key
		if label == "" {
			label = "未知"
		}
		fmt.Printf("    %-20s  %6d 题\n", label, m.Count)
	}
	fmt.Println(sep)
}

func printInspectQuestion(q *models.Question, sq *models.SubQuestion, qi, si, W int, full, showAI bool) {
	ansFlag := "❓"
	if strings.TrimSpace(sq.Answer) != "" {
		ansFlag = "✅"
	} else if strings.TrimSpace(sq.AIAnswer) != "" {
		ansFlag = "🤖"
	}

	disFlag := "❓"
	if strings.TrimSpace(sq.Discuss) != "" {
		disFlag = "✅"
	} else if strings.TrimSpace(sq.AIDiscuss) != "" {
		disFlag = "🤖"
	}

	aiTag := ""
	if sq.AIAnswer != "" || sq.AIDiscuss != "" {
		aiTag = " [AI]"
	}

	fmt.Println(strings.Repeat("─", W))
	fmt.Printf("  [%d-%d]  %s  %s  答案:%s  解析:%s%s\n",
		qi+1, si+1, q.Mode, q.Unit, ansFlag, disFlag, aiTag)

	if q.Stem != "" {
		stem := q.Stem
		if !full {
			stem = truncate(stem, 100)
		}
		fmt.Printf("  【题干】%s\n", stem)
	}

	text := sq.Text
	if text == "" {
		text = "(无题目文本)"
	} else if !full {
		text = truncate(text, 100)
	}
	fmt.Printf("  【题目】%s\n", text)

	// Options
	labels := "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
	optPrefix := regexp.MustCompile(`^[A-Za-z][.．、]\s*`)
	for i, opt := range sq.Options {
		opt = strings.TrimSpace(opt)
		if optPrefix.MatchString(opt) {
			fmt.Printf("         %s\n", opt)
		} else {
			lbl := "?"
			if i < len(labels) {
				lbl = string(labels[i])
			}
			fmt.Printf("         %s. %s\n", lbl, opt)
		}
	}

	// Answer
	effAns := sq.EffAnswer()
	if effAns != "" {
		src := ""
		conf := ""
		if sq.AnswerSource() == "ai" {
			src = " (AI)"
			conf = fmt.Sprintf("  [置信:%.2f  模型:%s]", sq.AIConfidence, sq.AIModel)
		}
		fmt.Printf("  【答案】%s%s%s\n", effAns, src, conf)
	}

	// Discuss
	effDis := sq.EffDiscuss()
	if effDis != "" {
		src := ""
		if sq.DiscussSource() == "ai" {
			src = " (AI)"
		}
		dis := effDis
		if !full {
			dis = truncate(dis, 150)
		}
		fmt.Printf("  【解析】%s%s\n", dis, src)
	}

	// Show AI raw output
	if showAI && (sq.AIAnswer != "" || sq.AIDiscuss != "") {
		sepLen := (W - 14) / 2
		fmt.Printf("  %s AI原始输出 %s\n", strings.Repeat("─", sepLen), strings.Repeat("─", sepLen))
		if sq.AIAnswer != "" && sq.AIAnswer != sq.EffAnswer() {
			fmt.Printf("  【AI答案】%s  [置信:%.2f  模型:%s]\n",
				sq.AIAnswer, sq.AIConfidence, sq.AIModel)
		}
		if sq.AIDiscuss != "" {
			dis := sq.AIDiscuss
			if !full {
				dis = truncate(dis, 150)
			}
			note := ""
			if strings.TrimSpace(sq.AIDiscuss) == strings.TrimSpace(sq.Discuss) {
				note = "（与官方解析相同）"
			}
			fmt.Printf("  【AI解析】%s%s\n", dis, note)
		}
	}
}

func truncate(s string, n int) string {
	s = strings.ReplaceAll(s, "\n", " ")
	s = strings.TrimSpace(s)
	runes := []rune(s)
	if len(runes) > n {
		return string(runes[:n]) + "…"
	}
	return s
}

func containsAny(s string, needles []string) bool {
	for _, n := range needles {
		if s == n {
			return true
		}
	}
	return false
}

func containsAnySubstring(s string, needles []string) bool {
	for _, n := range needles {
		if strings.Contains(s, n) {
			return true
		}
	}
	return false
}
